src/Aspire.Hosting/build/Aspire.Hosting.targets

```diff
@@ -13,17 +13,14 @@
   </ItemGroup>
 
   <!-- Generate the data structures for doing the codegen for service references -->
-  <Target Name="CreateServiceMetadataSources"
-          DependsOnTargets="FindReferenceAssembliesForReferences"
-          Inputs="@(ReferencePathWithRefAssemblies)"
-          Outputs="@(ReferencePathWithRefAssemblies)">
+  <Target Name="CreateServiceMetadataSources">
     <ItemGroup>
-      <ServiceMetadataSource Include="%(ReferencePathWithRefAssemblies.Identity)" Condition="%(ReferencePathWithRefAssemblies.ReferenceSourceTarget) == 'ProjectReference'">
-        <ClassName Condition="%(ReferencePathWithRefAssemblies.ReferenceSourceTarget) == 'ProjectReference' And %(ReferencePathWithRefAssemblies.ServiceNameOverride) == ''">$([System.IO.Path]::GetFileNameWithoutExtension(%(ReferencePathWithRefAssemblies.Identity)).Replace(".", "_").Replace("-","_").Replace(" ","_"))</ClassName>
-        <ClassName Condition="%(ReferencePathWithRefAssemblies.ReferenceSourceTarget) == 'ProjectReference' And %(ReferencePathWithRefAssemblies.ServiceNameOverride) != ''">$([System.String]::Copy(%(ReferencePathWithRefAssemblies.ServiceNameOverride)).Replace(".", "_").Replace("-","_").Replace(" ","_"))</ClassName>
-        <AssemblyName>$([System.IO.Path]::GetFileNameWithoutExtension(%(ReferencePathWithRefAssemblies.Identity)))</AssemblyName>
-        <ProjectPath>$([System.IO.Path]::GetFullPath(%(ReferencePathWithRefAssemblies.ProjectReferenceOriginalItemSpec)))</ProjectPath>
-        <Namespace>Projects</Namespace>
+      <_ServiceAppProject Include="@(ProjectReference->WithMetadataValue('_IsServiceApp', 'true'))" />
+
+      <ServiceMetadataSource Include="@(_ServiceAppProject)" Condition="'@(_ServiceAppProject)' != ''">
+        <ClassName Condition="%(_ServiceAppProject.ServiceNameOverride) == ''">$([System.IO.Path]::GetFileNameWithoutExtension(%(_ServiceAppProject.Identity)).Replace(".", "_").Replace("-","_").Replace(" ","_"))</ClassName>
```

Comment:
Should we update this from `ServiceName` to `ResourceName` now?

Reply:
Or maybe just `AspireResourceName` and that can be used as the opt-out too by setting it to empty?

Reply:
What about `ProjectsClassName`? Or even `ProjectsName`?

I don't like calling it "ResourceName" because that is what we call the string parameter that pass in to `AddProject`, which could be a different name than this name.

> and that can be used as the opt-out too by setting it to empty?

In MSBuild, I haven't seen setting something to empty and not setting it being different behavior. Usually checking for `== ''` means "it isn't set".

Reply:
For Items we do have the ['HasMetadata' item function](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-functions?view=vs-2022), but aside from this it can be hard to distinguish between missing and empty.

Reply:
I don't like setting this metadata to empty to mean something completely different than what the metadata is intended for.

* `<ProjectReference Include=.. />` - means this is a service app and don't reference it, use the default name for the Projects class name
* `<ProjectReference Include=.. AspireResourceName="foo" />` - means this is a service app and don't reference it, use "foo" for the Projects class name
* `<ProjectReference Include=.. AspireResourceName="" />` - means this isn't a service app and should be referenced

I do think we should rename `ServiceNameOverride`, but I don't know of a good name. If people think `AspireResourceName` is OK, I can rename it. I just have the reservations I listed above.